test/Interop/Cxx/objc-correctness/sret.swift

```diff
@@ -0,0 +1,72 @@
+// RUN: %empty-directory(%t)
+// RUN: split-file %s %t
+
+// RUN: %target-swift-emit-ir -I %t/Inputs -cxx-interoperability-mode=default %t/test.swift -target arm64-apple-macos12 | %FileCheck %s
+
+// REQUIRES: objc_interop
+
+//--- Inputs/header.h
+
+class S {
+public:
+  int a;
+  ~S();
+  S getS(int) const;
+  static S getSStatic(int);
+};
+
+S getS(int);
+
+@interface C
++(S)getS:(int)a;
+-(S)getS:(int)a;
+@end
+
+//--- Inputs/module.modulemap
+
+module SRet {
+  header "header.h"
+  requires cplusplus
+  export *
+}
+
+//--- test.swift
+
+import SRet
+
+func test(c : C, s : S) {
+  let _ : S = c.getS(1)
+  let _ : S = C.getS(1)
+  let _ : S = S()
+  let _ : S = getS(1)
+  let _ : S = s.getS(1)
+  let _ : S = S.getSStatic(1)
+}
+
+// CHECK: %[[TSO1SV:.*]] = type <{ %[[TS5INT32V:.*]] }>
+// CHECK: %[[TS5INT32V]] = type <{ i32 }>
+// CHECK: %[[SWIFT_OPAQUE:.*]] = type opaque
+// CHECK: %[[CLASS_S:.*]] = type { i32 }
+
+// CHECK: define hidden swiftcc void @"$s4testAA1c1sySo1CC_So1SVtF"(ptr %[[V0:.*]], ptr {{.*}}%[[V1:.*]])
+// CHECK: %[[V2:.*]] = alloca %[[TSO1SV]], align 4
+// CHECK: %[[V3:.*]] = alloca %[[TSO1SV]], align 4
+// CHECK: %[[V4:.*]] = alloca %[[TSO1SV]], align 4
+// CHECK: %[[V5:.*]] = alloca %[[TSO1SV]], align 4
+// CHECK: %[[V6:.*]] = alloca %[[TSO1SV]], align 4
+// CHECK: %[[V7:.*]] = alloca %[[TSO1SV]], align 4
+// CHECK: call void @llvm.lifetime.start.p0(i64 4, ptr %[[V2]])
+// CHECK: %[[V8:.*]] = load ptr, ptr @"\01L_selector(getS:)", align 8
+// CHECK: invoke void @objc_msgSend(ptr noalias nocapture sret(%[[SWIFT_OPAQUE]]) %[[V2]], ptr %[[V0]], ptr %[[V8]], i32 1)
```

Comment:
I think `ptr noalias nocapture sret(%[[TSO1SV]])` should be emitted here. Should `getSingleSILResultType` be called in `SignatureExpansion::addIndirectResult` instead of `getSILResultType`?

Reply:
the `sret` argument type should match the pointed to type of the argument. yes, `SWIFT_OPAQUE` is wrong.

Reply:
I suppose the type in `CHECK: invoke void @_Z4getSi(ptr noalias nocapture sret(%[[CLASS_S]]) %[[V5]], i32 1)` should be `TSO1SV` too instead of `CLASS_S`? But if I make that change, the called function's type is going to differ from the declared function clang generates, which is `declare void @_Z4getSi(ptr sret([[CLASS_S]])`, because the `sret` type is different. I don't think we can change the type of the `alloca`?

Reply:
Is `[[CLASS_S]]` the C type and `TSO1SV` the Swift imported version of the C type. I would think if the layout is equivalent we should be okay i.e it need not be literally the same type.

Reply:
Yes, that's correct.